src/ucp/proto/proto_common.c

```diff
@@ -692,6 +693,57 @@ void ucp_proto_common_zcopy_adjust_min_frag_always(ucp_request_t *req,
     }
 }
 
+ucs_status_t ucp_proto_request_init(ucp_request_t *req)
+{
+    ucp_ep_h ep         = req->send.ep;
+    ucp_worker_h worker = ep->worker;
+    ucp_worker_cfg_index_t rkey_cfg_index;
+    ucp_proto_select_t *proto_select;
+    size_t msg_length;
+    ucs_status_t status;
+
+    proto_select = ucp_proto_select_get(worker, ep->cfg_index,
+                                        req->send.proto_config->rkey_cfg_index,
+                                        &rkey_cfg_index);
+    if (proto_select == NULL) {
+        return UCS_OK;
+    }
+
+    msg_length = req->send.state.dt_iter.length;
+    if (ucp_proto_config_is_am(req->send.proto_config)) {
+        msg_length += req->send.msg_proto.am.header.length;
+    }
+
+    /* Select from protocol hash according to saved request parameters */
+    status = ucp_proto_request_lookup_proto(
+            worker, ep, req, proto_select, rkey_cfg_index,
+            &req->send.proto_config->select_param, msg_length);
+    if (status != UCS_OK) {
```

Comment:
here, just return status from ucp_proto_request_lookup_proto
only proto_reconfig.c should check status and return UCS_ERR_NO_RESOURCE, because it's a pending callback